File: src/main/java/tsingjyujing/geo/util/mathematical/VectorUtil.java
package tsingjyujing.geo.util.mathematical;


/**
 * @author tsingjyujing
 * @Mail tsingjyujing@163.com
 * @Telephone 182-2085-2215
 */
public class VectorUtil {

    public static double[] outerProduct(double[] v1, double[] v2) {
        assert (v1.length == 3);
        assert (v2.length == 3);
        double[] returnValue = {
                v1[1] * v2[2] - v1[2] * v2[1],
                -v1[0] * v2[2] + v1[2] * v2[0],
                +v1[0] * v2[1] - v1[1] * v2[0]
        };
        return returnValue;
    }

    public static void norm2Vector(double normLen, double[] vec) {
        double sumval = 0;
        for (double value : vec) {
            sumval += value * value;
        }
        double normFact = normLen / Math.sqrt(sumval);
        for (int i = 0; i < vec.length; i++) {
            vec[i] *= normFact;
        }
    }
}

File: src/main/scala/tsingjyujing/geo/util/mathematical/VectorUtil.scala
package tsingjyujing.geo.util.mathematical

object VectorUtil {
    def outerProduct(v1: Array[Double], v2: Array[Double]): Array[Double] = if ((v1.length == 3) && (v2.length == 3)){
        Array(
            0.0D + v1(1) * v2(2) - v1(2) * v2(1),
            0.0D - v1(0) * v2(2) + v1(2) * v2(0),
            0.0D + v1(0) * v2(1) - v1(1) * v2(0)
        )
    }else{
        throw new RuntimeException("Out product only allowed in 3D vector")
    }

    def norm2Vector(vec: Array[Double], normLen: Double = 1.0): Array[Double] = {
        val normFactor = normLen / math.sqrt(vec.map(x => x * x).sum)
        vec.map(x => x * normFactor)
    }
}
